Adjustments in nRF drivers and ext/hal/nordic to provide support for nRF9160

anangl opened the pull request:
**drivers: i2c: nrfx: Add support for TWIM2 and TWIM3**

Extend the i2c_nrfx_twim driver to support all TWIM instances
available on nRF9160.

**drivers: serial: nrfx: Add support for UARTE2 and UARTE3**

Extend the uart_nrfx_uarte driver to support all UARTE instances
available on nRF9160.

**dts: nrf: Add bindings for CLOCK, POWER, RTC, and TIMER peripherals**

Add dts bindings for nRF CLOCK, POWER, RTC, and TIMER peripherals.

**ext: hal: nordic: Add nrfx_config for nRF9160 SoC**

Add the header file with the nrfx drivers configuration suitable for
the nRF9160 SoC. This file contains also definitions of CMSIS symbols
(like NRF_CLOCK or NRF_SPIM0) for structures of peripheral registers.
These symbols are used in nrfx HALs and drivers, and for other nRF SoCs
they come from MDK files. For nRF9160 they need to be mapped to proper
addresses in Secure or Non-Secure domain, so we use the information
coming from DT to get it done automatically.

**dts: nrf9160: Add rtc0 node and replace nvmc node with alias**

Add missing node for instance 0 of the RTC peripheral and use an alias
to the flash-controller node instead of the nvmc node definition (there
is no need to have a separate node just to get the NVMC base address).

**drivers: nrf: Adjust clock_control and timer drivers for nRF9160**

Minor adjustments are done to the nRF clock_control and rtc_timer
drivers to make them usable on nRF9160 as well.

anangl committed:
drivers: i2c: nrfx: Add support for TWIM2 and TWIM3

Extend the i2c_nrfx_twim driver to support all TWIM instances
available on nRF9160.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: serial: nrfx: Add support for UARTE2 and UARTE3

Extend the uart_nrfx_uarte driver to support all UARTE instances
available on nRF9160.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
dts: nrf: Add bindings for CLOCK, POWER, RTC, and TIMER peripherals

Add dts bindings for nRF CLOCK, POWER, RTC, and TIMER peripherals.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
ext: hal: nordic: Add nrfx_config for nRF9160 SoC

Add the header file with the nrfx drivers configuration suitable for
the nRF9160 SoC. This file contains also definitions of CMSIS symbols
(like NRF_CLOCK or NRF_SPIM0) for structures of peripheral registers.
These symbols are used in nrfx HALs and drivers, and for other nRF SoCs
they come from MDK files. For nRF9160 they need to be mapped to proper
addresses in Secure or Non-Secure domain, so we use the information
coming from DT to get it done automatically.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
dts: nrf9160: Add rtc0 node and replace nvmc node with alias

Add missing node for instance 0 of the RTC peripheral and use an alias
to the flash-controller node instead of the nvmc node definition (there
is no need to have a separate node just to get the NVMC base address).

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: nrf: Adjust clock_control and timer drivers for nRF9160

Minor adjustments are done to the nRF clock_control and rtc_timer
drivers to make them usable on nRF9160 as well.
The arm_irq_vector_table test code is modified only because it uses
the function that has been renamed in the nrf_rtc_timer driver.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Removed one stray space.

ioannisg commented on dts/arm/nordic/nrf9160_common.dtsi:
```diff
@@ -161,6 +154,13 @@ gpio0: gpio@842500 {
 	status = "disabled";
 };
 
+rtc0: rtc@14000 {
```
Why was this needed?

ioannisg replied:
Also, no alias is defined for this one....

anangl replied:
> Why was this needed?

To get the base address of this peripheral in [nrfx_config_nrf9160.h](https://github.com/zephyrproject-rtos/zephyr/pull/12531/files#diff-7c92595657f7db412148884cfa771899R80).

> Also, no alias is defined for this one....

Right. This is missing. I'll add the alias.

anangl replied:
Done.

ioannisg commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -11,12 +11,16 @@
 #endif
 
+#if defined(CONFIG_SOC_SERIES_NRF91X)
+#define POWER_CLOCK_IRQn  CLOCK_POWER_IRQn
```
Great. Do you think this should be done in this PR?

anangl replied:
I can do it here. In a separate commit, to make it more visible.

anangl replied:
Done.

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -254,6 +259,7 @@ static int _k32src_start(struct device *dev, clock_control_subsys_t sub_system)
 			__ASSERT_NO_MSG(err == -EINPROGRESS);
 		}
 	}
+#endif /* NRF_CLOCK_HAS_CALIBRATION */
```
Because when the corresponding `#if` is only a couple of lines above and it can be easily identified without the help of such comment, there is no need to add one.
In places like this:
```
#if NRF_CLOCK_HAS_CALIBRATION
	u8_t ctto, done;
	struct device *dev = arg;
#endif
#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
	bool usb_detected, usb_pwr_rdy, usb_removed;
#endif
```
adding of such comments after `#endif` does not necessarily improve the code readability:
```
#if NRF_CLOCK_HAS_CALIBRATION
	u8_t ctto, done;
	struct device *dev = arg;
#endif /* NRF_CLOCK_HAS_CALIBRATION */
#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
	bool usb_detected, usb_pwr_rdy, usb_removed;
#endif /* defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840) */
```

anangl commented:
I am not sure how should I fix the CI [issue](https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/31284/1/console).
BT simulator tests cannot be built because of undeclared symbol `NRF_CLOCK_HFCLK_HIGH_ACCURACY`. The problem is that the symbol is declared [in _<nrf_clock.h>_](https://github.com/zephyrproject-rtos/zephyr/blob/master/ext/hal/nordic/nrfx/hal/nrf_clock.h#L119) but the simulator uses its own, [stripped-down version of this file](https://github.com/BabbleSim/ext_NRF52_hw_models/blob/master/src/nrfx/hal/nrf_clock.h) and hence does not have it. @aescolar Could you advise?

ioannisg commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -49,7 +49,7 @@ static u32_t counter(void)
  * symbol.
  */
-void rtc1_nrf5_isr(void *arg)
+void rtc1_nrf_isr(void *arg)
```
could this function be declared as static?

anangl replied:
I guess not. Because the arm_irq_vector_table [apparently uses it](https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/kernel/arm_irq_vector_table/src/arm_irq_vector_table.c#L123). I need to rename it in this test as well.

anangl replied:
Done (renaming of the function in the test).

anangl committed:
drivers: clock_control: nrf: Use IRQ number symbol generated from DT

Use the symbol generated from DT so that it is common for nRF9160
and nRF5 family SoCs. To avoid artificial renaming of CLOCK_POWER_IRQn
to POWER_CLOCK_IRQn.

For nRF5 family SoCs clock nodes were not defined so far, thus they are
added so that the proper DT symbol is generated for them as well.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented on ext/hal/nordic/nrfx_config_nrf9160.h:
```diff
@@ -0,0 +1,2147 @@
+#ifndef NRFX_CONFIG_NRF9160_H__
+#define NRFX_CONFIG_NRF9160_H__
+
+// The definitions below
```
Done.

ioannisg commented:
Belongs to partent issue #12199

carlescufi reviewed (approved):
LGTM

ioannisg commented:
Somehow the CI needs to pass, though

lemrey reviewed (approved):
LGTM, have done a quick check locally.

ioannisg commented:
@anangl I believe some of the failures are due to the use of DT_NORDIC_NRF_CLOCK_0_IRQ_0, throughout the clock-control driver. There is no DTS info for clock-control yet for nrf51,52, so these defines are not auto-generated, as far as I understand

aescolar commented:
> BT simulator tests cannot be built because of undeclared symbol NRF_CLOCK_HFCLK_HIGH_ACCURACY. The problem is that the symbol is declared in <nrf_clock.h> but the simulator uses its own, stripped-down version of this file and hence does not have it. @aescolar Could you advise?

@anangl : (The nRF52 HW models use their own HAL version) Should be easy to fix, I should be able to push a patch tomorrow to align it with the latest nRFx HAL.
Note that there is also a DT define issue (both in the nrf52_bsim and in a few real boards)

anangl commented:
> Note that there is also a DT define issue (both in the nrf52_bsim and in a few real boards)

Right, my bad. Fixed.

anangl commented:
@ioannisg
> @anangl I believe some of the failures are due to the use of DT_NORDIC_NRF_CLOCK_0_IRQ_0, throughout the clock-control driver. There is no DTS info for clock-control yet for nrf51,52, so these defines are not auto-generated, as far as I understand

You're right. I fixed that by adding clock nodes for nRF5 family SoCs. Please, take a look.

anangl commented:
> @ioannisg
> 
> > @anangl I believe some of the failures are due to the use of DT_NORDIC_NRF_CLOCK_0_IRQ_0, throughout the clock-control driver. There is no DTS info for clock-control yet for nrf51,52, so these defines are not auto-generated, as far as I understand
> 
> You're right. I fixed that by adding clock nodes for nRF5 family SoCs. Please, [take a look](https://github.com/zephyrproject-rtos/zephyr/compare/f2f2961bac3541d056b53f3aa44c325019cf1d6a..f219a118b6e90d3990bd9590fdfc24196111378c).

ioannisg commented:
recheck

anangl commented:
> recheck

It will still fail. We must wait until @aescolar updates _nrf_clock.h_ in [ext_NRF52_hw_models](https://github.com/BabbleSim/ext_NRF52_hw_models/blob/master/src/nrfx/hal/nrf_clock.h).

anangl committed:
dts: nrf9160: Add nodes for PWM peripherals

Add DT nodes for PWM peripherals available in nRF9160.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Added adjustments needed to utilize PWM peripherals.

ioannisg commented:
> 
> 
> > BT simulator tests cannot be built because of undeclared symbol NRF_CLOCK_HFCLK_HIGH_ACCURACY. The problem is that the symbol is declared in <nrf_clock.h> but the simulator uses its own, stripped-down version of this file and hence does not have it. @aescolar Could you advise?
> 
> @anangl : (The nRF52 HW models use their own HAL version) Should be easy to fix, I should be able to push a patch tomorrow to align it with the latest nRFx HAL.
> Note that there is also a DT define issue (both in the nrf52_bsim and in a few real boards)

@aescolar the DT issue is (hopefully) addressed.

ioannisg commented on dts/arm/nordic/nrf9160_common.dtsi:
```diff
@@ -152,6 +145,38 @@ spi3: spi@b000 {
 	label = "SPI_3";
 };
 
+pwm0: pwm@21000 {
```
This has only be added for nRF9160, right?

anangl replied:
I am not sure if I correctly understood the question.
These nodes were missing for nRF9160. nRF5 family has them already defined (and all the other stuff, like bindings, is already in place).

ioannisg replied:
Ah right - sorry, I did not check thoroughly.

anangl committed:
drivers: pwm: nrf: Correct dependency of PWM_NRFX Kconfig option

Correct it so that it is possible to use PWMs on nRF9160 as well,
not only on nRF5 family SoCs.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Forgot to add one PWM related commit in the last update.

aescolar reviewed (commented):
If you are changing the POWER_CLOCK_IRQn , seems you forgot nrf_power_clock.c : 484

anangl commented:
> If you are changing the POWER_CLOCK_IRQn , seems you forgot nrf_power_clock.c : 484

How could that happen? I was sure I had replaced this one. Anyway, fixed.
Thanks a lot for spotting this!

aescolar commented:
> If you are changing the POWER_CLOCK_IRQn 

@anangl : For curiosity, what is the motivation to change from using that MDK type into using the DTS macro?

anangl commented:
> > If you are changing the POWER_CLOCK_IRQn
> 
> @anangl : For curiosity, what is the motivation to change from using that MDK type into using the DTS macro?

@aescolar For nRF9160 this enum value is named CLOCK_POWER_IRQn, and we wanted to avoid using renaming like [this](
 https://github.com/zephyrproject-rtos/zephyr/compare/eda56e292d9d50d031fdd01fb20f123fb3f25d5c..f2f2961bac3541d056b53f3aa44c325019cf1d6a#diff-3c5c76aad21e56267c253f5caa8c09e8L20) (this was the initially proposed change in this PR). Another alternative could be `NRFX_IRQ_NUMBER_GET(NRF_CLOCK)`. I thought the solution with DT symbol would be better.

aescolar commented:
> For nRF9160 this enum value is named CLOCK_POWER_IRQn

Oh well, I guess the issue then was not being coherent in the MDK side..

aescolar commented:
recheck

ioannisg commented:
This one is good to go, IMO; nrfx_config_nrf9160.h has the standard ext/hal/nordic license.
